src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -952,7 +1121,91 @@ internal static object ComObjectForInterface(IntPtr externalComObject)
+            return s_globalInstanceForTrackerSupport.GetOrCreateComInterfaceForObject(obj, CreateComInterfaceFlags.TrackerSupport);
+        }
+
+        internal static void ReleaseExternalObjectsFromCurrentThread()
+        {
+            if (s_globalInstanceForTrackerSupport == null)
+            {
+                throw new NotSupportedException(SR.InvalidOperation_ComInteropRequireComWrapperInstance);
+            }
+
+            IntPtr contextToken = GetContextToken();
+
+            List<object> objects = new List<object>();
+            foreach (GCHandle weakNativeObjectWrapperHandle in s_nativeObjectWrapperCache)
+            {
+                NativeObjectWrapper? nativeObjectWrapper = Unsafe.As<NativeObjectWrapper?>(weakNativeObjectWrapperHandle.Target);
+                if (nativeObjectWrapper != null &&
+                    nativeObjectWrapper._fromTrackerRuntime &&
+                    nativeObjectWrapper._contextToken == contextToken)
+                {
+                    objects.Add(nativeObjectWrapper._proxyHandle.Target);
+
+                    // Separate the wrapper from the tracker runtime prior to
+                    // passing them.
+                    nativeObjectWrapper.DisconnectTracker();
+                }
+            }
+
+            s_globalInstanceForTrackerSupport.ReleaseObjects(objects);
+        }
+
+        internal static unsafe void WalkExternalTrackerObjects()
+        {
+            bool walkFailed = false;
+
+            foreach (GCHandle weakNativeObjectWrapperHandle in s_nativeObjectWrapperCache)
+            {
+                NativeObjectWrapper? nativeObjectWrapper = Unsafe.As<NativeObjectWrapper?>(weakNativeObjectWrapperHandle.Target);
+                if (nativeObjectWrapper != null &&
+                    nativeObjectWrapper.TrackerObject != IntPtr.Zero)
+                {
+                    FindReferenceTargetsCallback.s_currentRootObjectHandle = nativeObjectWrapper._proxyHandle;
+                    if (IReferenceTracker.FindTrackerTargets(nativeObjectWrapper.TrackerObject, TrackerObjectManager.s_findReferencesTargetCallback) != 0)
+                    {
+                        walkFailed = true;
+                        FindReferenceTargetsCallback.s_currentRootObjectHandle = default;
+                        break;
+                    }
+                    FindReferenceTargetsCallback.s_currentRootObjectHandle = default;
+                }
+            }
+
+            // Report whether walking failed or not.
+            if (walkFailed)
+            {
+                TrackerObjectManager.s_isGlobalPeggingOn = true;
+            }
+            IReferenceTrackerManager.FindTrackerTargetsCompleted(TrackerObjectManager.s_trackerManager, walkFailed);
+        }
+
+        internal static void DetachNonPromotedObjects()
+        {
+            for (int idx = 0; idx < s_nativeObjectWrapperCache.Count; idx++)
```

Comment:
How come this isn't also just using a `foreach` loop like in `WalkExternalTrackerObjects`? Just trying to understand this better — does this method expect that the wrapper cache list will be mutated across iterations?

Reply:
I was using a different collection before and that was doing an allocation when using the enumerator which causes issues for GC callbacks such as this.  So I had addressed it by doing that, but it seems to no longer be the issue with the current collection I am using.  Updated to use foreach again.